app/controllers/application_controller.rb

```diff
@@ -3,8 +3,139 @@
 class ApplicationController < Sinatra::Base
 
   configure do
+    enable :sessions
+    set :session_secret, 'billy'
     set :public_folder, 'public'
     set :views, 'app/views'
   end
 
-end
\ No newline at end of file
+  get '/' do
+    erb :index
+  end
+
+  get '/signup' do
+    if session[:id]
+      redirect '/tweets'
+    end
+      erb :signup
+  end
+
+  post '/signup' do
+    @user = User.find_or_create_by(username: params['username'], email: params['email'], password: params['password'])
+    if @user.save
+      session[:id] = @user.id
+      redirect '/tweets'
+    else
+      redirect '/signup'
+    end
+  end
+
+  get '/login' do
+    if session[:id] == nil
+      erb :login
+    else
+      redirect '/tweets'
+    end
+  end
+
+  post '/login' do
+    @user = User.find_by(username: params['username'], password: params['password'])
+    session[:id] = @user.id
+    redirect '/tweets'
+  end
+
+  get '/tweets' do
+    if is_logged_in?(session)
+      @user = current_user(session)
+      @tweets = Tweet.all
+      erb :tweets
+    else
+      redirect '/login'
+    end
+
+  end
+
+  get '/logout' do
+    if is_logged_in?(session)
+      session.clear
+      redirect '/login'
+    else
+      redirect '/'
+    end
+  end
+
+  get '/users/:slug' do
+    @user = User.find_by_slug(params(:slug))
+    erb :show
+  end
+
+  get '/tweets/new' do
+    if is_logged_in?(session)
+      @user = current_user(session)
+      erb :newtweet
+    else
+      redirect '/login'
+    end
+  end
+
+  post '/newtweet' do
+    if params["content"] != ""
+      @newtweet = Tweet.new(content: params['content'])
+      @newtweet.user_id = current_user(session).id
+      @newtweet.save
+      redirect "/tweets/#{@newtweet.id}"
+    else
+      redirect '/tweets/new'
+    end
+  end
+
+  get '/tweets/:id' do
+    if is_logged_in?(session)
+      @tweet = Tweet.find(params["id"])
+      erb :showtweet
+    else
+      redirect '/login'
+    end
+  end
+
+  get '/tweets/:id/edit' do
+    if is_logged_in?(session)
+      @tweet = Tweet.find(params["id"])
+      erb :edittweet
+    else
+      redirect '/login'
+    end
+  end
+
+  post '/tweets/:id/edit' do
+    @tweet = Tweet.find(params["id"])
+    @tweet.content = params["content"]
+    @tweet.save
+  end
+
+  delete '/tweets/:id/delete' do
+    @tweet = Tweet.find(params["id"])
+    if is_logged_in?(session) && current_user(session).id == @tweet.user_id
```

Comment:
Prefer to find tweets that are already only associated to the `current_user` than to manually check the `user_id` foreign key of the tweet against the `current_user.id`

```ruby
tweet = current_user.tweets.find_by(:id => params[:id])
if tweet && tweet.destroy
  redirect "/tweets"
else
  redirect "/tweets/#{tweet.id}"
end
```

We're using the association collection method provided by [`has_many`](http://api.rubyonrails.org/classes/ActiveRecord/Associations/ClassMethods.html#method-i-has_many) to query the user's tweets. 

Note the usage of [`find_by`](http://api.rubyonrails.org/classes/ActiveRecord/FinderMethods.html#method-i-find-by) vs [`find`](http://api.rubyonrails.org/classes/ActiveRecord/FinderMethods.html#method-i-find). Had we used `find` here, ActiveRecord would raise an error in the event that we could not find a tweet associated with this user. `find_by` will allow this action to fail gracefully, simply returning nil if the user is trying to delete a tweet that they do not have access to.